### PR TITLE
Understanding c++ templates

### LINKED ISSUE #1632: Understanding c++ templates
Be sure to visit our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for tools, resources, and example articles to go over.

**NOTE: We tend to stray away or tend not to publish reviews/comparisons of commercial product offerings.**

### Proposed title of article
Understanding c++ templates

### Introduction paragraph(s):
A template is a method of writing a single function or class for a family of related functions or classes. In plain terminology, you can declare a single function or a class to work with different data types using templates.  This helps the developer avoid redundant code in a program. In the broader codebase, we use templates for code reusability and software flexibility. This article will take a look at templates and how to use them for generic programming in c++. 

### Key takeaways: 
1. How templates work and when to use them
2. Class and function templates
3. Variadic function templates
4. How to create a class and function template using intuitive programs
5. Advantages of templates

### References: 
N/A

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by Dawe-7**
### Article checklist
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

This fixes #1632

### CODE REVIEW

File: articles/understanding-c++-templates/index.md [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+- `Function templates`
+
+### How do templates work?
+Templates expand at compile time like macros. The difference is that the compiler does a type check before expanding the template. The principle is basic, source code only contains function/class, but many copies of the same function/class can be included in compiled code.
+
+### Class templates
+A class template is a template that enables the developer to operate with generic data types. This enables the class to be used on several different data types as per the specifications without the need for each type to be re-written.
+Quite often, you'll need a class implementation that is the same for all classes except for the data types used. For each data type, you'd have to declare a separate class or build many member variables and functions within a single class.
+
+This will bloat the code base and will be difficult to manage because of a change in one class that will need change on all classes/functions. But, class templates allow you to reuse the same code for all data types. Like normal arguments, we can transfer more than one type of data to templates as arguments if we need to.
```
Comment (adrianmurage) [inline articles/understanding-c++-templates/index.md:20]:
```suggestion
The declaration of separate classes that do the same thing for different data types will bloat the code base and make it difficult to manage. Any changes made in one class will need to be reflected in all the similar classes. 

Class templates allow you to reuse the same code for all data types. Like normal arguments, we can transfer more than one type of data to templates as arguments if we need to.
```

File: articles/understanding-c++-templates/index.md [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+Floating-point numbers: 4.5, 9.3
+Characters: a, m
+
+The function template `Large()` is declared in the above program, which accepts two `d1` and `d2` arguments of data type `T`. `T` indicates any type can be an argument. The larger of the two arguments is returned by the `Large() function` using a conditional operator.
+
+Variables of three distinct data types are declared within the `main()` function. The variables will then be passed as normal functions to the `Large()` function prototype. When we pass an integer to the function template during runtime, the compiler knows that the `Large()` function must accept `int` arguments.
+
+Likewise, when `float` and `char` data are passed, the `Large()` function is generated because it recognizes the argument data types. This way, three identical normal functions were replaced by using a single function prototype.
+
+#### Variadic function template
```
Comment (adrianmurage) [inline articles/understanding-c++-templates/index.md:210]:
```suggestion
### Variadic function template
```

File: articles/understanding-c++-templates/index.md [hunk 3]
```diff
@@ -0,0 +1,321 @@
+### Introduction
```
Comment (adrianmurage) [inline articles/understanding-c++-templates/index.md:1]:
```suggestion
## Introduction
```

File: articles/understanding-c++-templates/index.md [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+A template is a way of writing a single function or class that can be reused by a group of similar functions or classes. In plain terminology, you can use templates to declare a single function or class that works with different data types. This article will take a look at templates and how to use them for generic programming in c++.
+
+### Prerequisites
```
Comment (adrianmurage) [inline articles/understanding-c++-templates/index.md:4]:
```suggestion
## Prerequisites
```

File: articles/understanding-c++-templates/index.md [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+To follow this article, you’ll need to have:
+- A [Codeblocks IDE](http://www.codeblocks.org/downloads) to run the code.
+- An understanding of the C++ language.
+
+We can use the concept of templates in two different ways:
+- `Class templates`
+- `Function templates`
+
+### How do templates work?
```
Comment (adrianmurage) [inline articles/understanding-c++-templates/index.md:13]:
```suggestion
## How do templates work?
```

File: articles/understanding-c++-templates/index.md [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+Templates expand at compile time like macros. The difference is that the compiler does a type check before expanding the template. The principle is basic, source code only contains function/class, but many copies of the same function/class can be included in compiled code.
+
+### Class templates
```
Comment (adrianmurage) [inline articles/understanding-c++-templates/index.md:16]:
```suggestion
## Class templates
```

File: articles/understanding-c++-templates/index.md [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+A class template is a template that enables the developer to operate with generic data types. This enables the class to be used on several different data types as per the specifications without the need for each type to be re-written.
+Quite often, you'll need a class implementation that is the same for all classes except for the data types used. For each data type, you'd have to declare a separate class or build many member variables and functions within a single class.
+
+This will bloat the code base and will be difficult to manage because of a change in one class that will need change on all classes/functions. But, class templates allow you to reuse the same code for all data types. Like normal arguments, we can transfer more than one type of data to templates as arguments if we need to.
+
+#### Declaration of a class template
```
Comment (adrianmurage) [inline articles/understanding-c++-templates/index.md:22]:
```suggestion
### Declaration of a class template
```

File: articles/understanding-c++-templates/index.md [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+Let’s break down the syntax as shown below:
+
+- `template <class T>`- It is the keyword to declare a template.
+- `class` - It is a user-defined type.
+- `class_name` - It is the class name for the class.
+- `T` - This is the argument for the template that is a placeholder for the data type which will be used.
+- `var` - It is a variable of the class of the generic data type.
+- ` T functionName(T arg)` - It is a function of the generic data type.
+
+#### How to create a class template object?
```
Comment (adrianmurage) [inline articles/understanding-c++-templates/index.md:44]:
```suggestion
### How to create a class template object?
```

File: articles/understanding-c++-templates/index.md [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+
+To build the function declaration outside the class template, we first declare the generic return type `T`  and then the class name `sample`. Right after the class name, we add `<T>` to tell the compiler that the function’s template  argument is the same one that we are using for the class.
+
+We compare the two digits within the function definition. We have created a class object using the same format we saw before the program:
+
+`  sample <int> big(120, 80);`
+
+Every time we are using the class template, it is required to tell it what data type we are substituting for `T`. In this instance, we are substituting `T` for `int`. Afterward, we created the object `big` and passed two integer numbers. We finish the program by calling the function `T larger()`.
+
+#### Advantages of class templates
```
Comment (adrianmurage) [inline articles/understanding-c++-templates/index.md:115]:
```suggestion
### Advantages of class templates
```

File: articles/understanding-c++-templates/index.md [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+The main advantages of using a class template are:
+- You need to define only one class that will operate with various types of data.
+- Only those data types for which the template has been used in the application produce instances of this template class at compile time.
+- As we have already established, the creation and use of a template class will cut the development efforts and lines of code. It will also decrease the difficulty and time of debugging the software for any problems because you only deal with 1 class.
+
+### Function templates
```
Comment (adrianmurage) [inline articles/understanding-c++-templates/index.md:121]:
```suggestion
## Function templates
```

File: articles/understanding-c++-templates/index.md [hunk 11]
```diff
@@ -0,0 +1,321 @@
...
+A [function template](https://bcastudyguide.wordpress.com/unit-4-generic-function) acts similarly to a normal function, with one key difference. A single function template can [interact with different data types at once](https://bcastudyguide.wordpress.com/unit-4-generic-function), but with a single normal function, only one set of data types can work.
+
+Examples of OOP's polymorphism feature are both function overloading and function templates.  If you need to perform similar operations for two or more data types, you use overloading functions to create two functions. When several functions perform identical operations, templates are used. Using function templates, but, would be a safer method since you can do the same task with less and more maintainable code
+
+#### Declaration of a function template?
```
Comment (adrianmurage) [inline articles/understanding-c++-templates/index.md:126]:
```suggestion
### Declaration of a function template?
```

File: articles/understanding-c++-templates/index.md [hunk 12]
```diff
@@ -0,0 +1,321 @@
...
+display("I will display\n");
+```
+
+```c++
+cout<< "I will display\n" <<endl ;
+display();
+```
+The compiler at this point searches for a function overload that fits the empty function, that is, a function that takes no arguments. All functions with one or more arguments are matched to the variadic function, while the empty are matched to the empty functions.
+
+### Conclusion
```
Comment (adrianmurage) [inline articles/understanding-c++-templates/index.md:318]:
```suggestion
## Conclusion 
```

**Review by adrianmurage — CHANGES_REQUESTED** [review]
@Dawe-7 

I left a few sentence structure suggestions. 
I also noticed that you were using H4 headings. According to the pre-review checklist, the preferred headings are H3, H2, and H1.

Aside from that this was a great read!
Once you've made the suggested changes @ me and I'll mark the article as ready for final review.

**Comment by Dawe-7** [comment]
Hello, @adrianmurage I've made the requested changes. I hope its good to go, thanks.

**Comment by adrianmurage** [comment]
> Hello, @adrianmurage I've made the requested changes. I hope its good to go, thanks.

Yup it is @Dawe-7 
Great article 👍🏽 

@hectorkambow this article is ready for final review

**Comment by hectorkambow** [comment]
great work gentlemen - be sure to go over any final edits 👍  @Dawe-7 

cc @adrianmurage

**Pull request merged**